File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 9/14/2025, 9:19:07 AM
class Solution:
    
    def isValid(self, s: str) -> bool:
        stack=[]
        map={'}':'{',')':'(',']':'['}
        for i in s:
            if i in map:
                #take the top element if stack is not empty otherwise assign duplicate value
                top_ele=stack.pop() if stack else '#'
                if top_ele!=map[i]:
                    return False
            else:
                stack.append(i)
        #not stack will return stack is empty or not ie true or false
        return not stack

File: 20-ValidParentheses/20-ValidParentheses.c
// Last updated: 9/14/2025, 9:19:03 AM
bool isValid(char * s){
    if(strlen(s)%2==1){
        return false;
    }
    char stack[10000];
    int top=-1;
    for(int i=0;s[i]!='\0';i++){
            switch(s[i]){
                case '(':
                case '{':
                case '[':
                     stack[++top]=s[i];
                     break;
                case ')':{
                      if(top<0 || stack[top]!='('){
                          return false;
                      }else{
                          top--;
                      }
                      break;
                    }
                case ']':{
                      if(top<0 || stack[top]!='['){
                          return false;
                      }else{
                          top--;
                      }
                      break;
                    }
                case '}':{
                      if(top<0 || stack[top]!='{'){
                          return false;
                      }else{
                          top--;
                      }
                      break;
                    }
                default:
                      return false;
            }
    }
      return top==-1;
}